src/platform/plugins/shared/workflows_execution_engine/server/step/http_step/http_step_impl.ts

```diff
@@ -0,0 +1,192 @@
+import type { UrlValidator } from '../../lib/url_validator';
+import { WorkflowTemplatingEngine } from '../../templating_engine';
+import { parseDuration } from '../../utils/parse-duration/parse-duration';
+import type { WorkflowContextManager } from '../../workflow_context_manager/workflow_context_manager';
+import type { WorkflowExecutionRuntimeManager } from '../../workflow_context_manager/workflow_execution_runtime_manager';
+import type { IWorkflowEventLogger } from '../../workflow_event_logger/workflow_event_logger';
+import type { StepImplementation } from '../step_base';
+
+type HttpHeaders = Record<string, string | number | boolean>;
+
+export class HttpStepImpl implements StepImplementation {
+  private templatingEngine: WorkflowTemplatingEngine;
+
+  constructor(
+    private node: HttpGraphNode,
+    private contextManager: WorkflowContextManager,
+    private workflowRuntime: WorkflowExecutionRuntimeManager,
+    private workflowLogger: IWorkflowEventLogger,
+    private urlValidator: UrlValidator
+  ) {
+    this.templatingEngine = new WorkflowTemplatingEngine();
+  }
+
+  public getInput() {
+    const context = this.contextManager.getContext();
+    const {
+      url,
+      method = 'GET',
+      headers = {},
+      body,
+      timeout = '30s',
+    } = this.node.configuration.with;
+
+    return {
+      url: typeof url === 'string' ? this.templatingEngine.render(url, context) : url,
+      method,
+      headers: this.renderHeaders(headers, context),
+      body: this.renderBody(body, context),
+      timeout: parseDuration(timeout),
+    };
+  }
+
+  private renderHeaders(headers: HttpHeaders, context: any): HttpHeaders {
+    return Object.entries(headers).reduce((acc, [key, value]) => {
+      acc[key] = typeof value === 'string' ? this.templatingEngine.render(value, context) : value;
+      return acc;
+    }, {} as HttpHeaders);
+  }
+
+  private renderBody(body: any, context: any): any {
+    if (typeof body === 'string') {
+      return this.templatingEngine.render(body, context);
+    }
+    if (body && typeof body === 'object') {
+      return this.renderObjectTemplate(body, context);
+    }
+    return body;
+  }
+
+  /**
+   * Recursively render the object template.
+   * @param obj - The object to render.
+   * @param context - The context to use for rendering.
+   * @returns The rendered object.
+   */
+  private renderObjectTemplate(obj: any, context: any): any {
+    if (Array.isArray(obj)) {
+      return obj.map((item) => this.renderObjectTemplate(item, context));
+    }
+    if (obj && typeof obj === 'object') {
+      return Object.entries(obj).reduce((acc, [key, value]) => {
+        acc[key] = this.renderObjectTemplate(value, context);
+        return acc;
+      }, {} as any);
+    }
+    if (typeof obj === 'string') {
+      return this.templatingEngine.render(obj, context);
+    }
+    return obj;
+  }
+
+  async run(): Promise<void> {
```

Comment:
you should implement `_run()` method, not the `run()`, as per comment in step_base.ts 
https://github.com/elastic/kibana/blob/main/src/platform/plugins/shared/workflows_execution_engine/server/step/step_base.ts#L91-L92

this way you won't need to duplicate logic of using setStepResult, handleFailure, etc

Reply:
good catch!